### ISSUE TITLE
New gstreamer sink leads to not working video presentations

### ISSUE THREAD

**Issue by AndreasBilke** [pdfpc/pdfpc#240]
With 84a9e337845f9229228ef4274895b398b5d07041 video presentations does not work properly (I use our pdf video example from the website).

If I come for the first time to a slide with a video, all following slide changes results in a overlaying video on the presenter and presentation view. Pdfpc reacts to slide change requests and on the next slide on presenter view I see the `next` slide, but the main view still shows the old video.

**Comment by AndreasBilke** [comment]
For my system: I use arch linux with every library at the latest version.

**Comment by phillipberndt** [comment]
I can reproduce this on my Desktop :( But I believe that your description of the problem is wrong: What you see isn't the previous video, but the actual one from slide 3 - only that it is displayed full-screen instead of within the prescribed bounds. At least, that's what happens here.

**Comment by phillipberndt** [comment]
cc @severin-lemaignan 
ref #197

**Comment by phillipberndt** [comment]
According to the [GstVideoOverlay documentation](https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-base-libs/html/GstVideoOverlay.html#gst-video-overlay-expose), it might be the case that the gl sink doesn't support the `gst_video_overlay_set_render_rectangle` method. Unfortunately, vala doesn't support the boolean return value on my system yet, so I can't test this, and I don't understand the [gstglimagesink source code](https://github.com/ystreet/gst-plugins-gl/blob/master/gst/gl/gstglimagesink.c) well enough yet to judge whether it does..

**Comment by severin-lemaignan** [comment]
I confirm the issue. @phillipberndt is indeed right: the bounds are not enforced, and the video always plays fullscreen.

**Comment by severin-lemaignan** [comment]
...let's ask the gstreamer devs... #gstreamer on freenode...

**Comment by severin-lemaignan** [comment]
@phillipberndt according the the gstreamer documentation you've linked, another option would be to create a new subwindow at the right position...

**Comment by phillipberndt** [comment]
We'll probably run into issues when we eventually migrate to Gtk 4 with that option, because afaik, the developers try to get rid of subwindows, and I can imagine that they'd remove the API to create them entirely to enforce that policy. But I agree that this is a viable solution for the time being.

**Comment by severin-lemaignan** [comment]
Well, interestingly enough, if you look at the presenter window, the movie does not play 'full screen'. It only covers the `View.Pdf` corresponding to the current slide. [`View.Pdf` is a `Gtk.DrawingArea`](https://github.com/pdfpc/pdfpc/blob/master/src/classes/view/pdf.vala#L35). So we could essentially create one new `Gtk.DrawingArea` per movie, properly positioned and size on top of the slide view.

**Comment by severin-lemaignan** [comment]
I've tried first to add a `Gtk.Fixed` widget to `View.Pdf` (to be able to add and position sub-widgets onto PDF views) and then to modify [`PresentationController.overlay_pos`](https://github.com/pdfpc/pdfpc/blob/master/src/classes/presentation_controller.vala#L1273) to create a new `Gtk.DrawingArea` for each movie with this code:
```vala
              Gtk.DrawingArea moviearea = new Gtk.DrawingArea();
              moviearea.set_size_request(rect.width*gdk_scale, rect.height*gdk_scale);
              view.fixed_layout.put(moviearea, rect.x*gdk_scale, rect.y*gdk_scale);
   
              return (uint*) ((Gdk.X11.Window) moviearea.get_window()).get_xid();
```
This currently segfaults:
```
Gtk-CRITICAL **: gtk_widget_realize: assertion 'widget->priv->anchored || GTK_IS_INVISIBLE (widget)' failed
```

I guess I did not do what was needed for the `Gtk.Fixed` to be actually added to the `View.Pdf`... anyway, my very limited knowledge of Gtk3 + tiredness makes it hard for me to fix it for now... help welcome!

**Comment by AndreasBilke** [comment]
@severin-lemaignan there is already a Gtk.Fixed layout in the fullscreen
class. It is used for used highlighting dots. Could you just attach your
widget to the existing Gtk.Fixed container?

**Comment by severin-lemaignan** [comment]
Interestingly, the gstreamer devs (ndufresne) told me that `glimagesink` is supposed to implement `set_render_rectangle`. I've opened a bug upstream: https://bugzilla.gnome.org/show_bug.cgi?id=782959

**Comment by phillipberndt** [comment]
I have implemented the fix in the [bug240-subwindow branch in my repo](https://github.com/phillipberndt/pdfpc/tree/bug240-subwindow): If the `glimagesink` is used, the program now creates a `Gtk.EventBox` (which comes with its own XWindow too, but doesn't draw a black background by default) and uses this subwindow's XID for drawing. The fix is a bit more complex than I imagined, because it has to deal with deletion/management of the subwindows and mouse event processing as well. I can imagine cleaner solutions that what I wrote if we refactor the APIs a bit, I went for "minimal-invasive" here.

**Comment by AndreasBilke** [comment]
@phillipberndt Your branch fixes the problem described above, but I have new problems. Sometimes everything works fine but I got multiple times a flickering video (either on presenter or on presentation but not both at the same time) with no running video (but with running video on the other screen).

**Comment by AndreasBilke** [comment]
Before you go here further, can we combine the efforts to fix #210 ? I actually don't understand this gstreamer chain but it seems that the splitting of the video source to two sinks leads to a situation where the "smallest" is chosen.

**Comment by phillipberndt** [comment]
Hm, I cannot reproduce the flickering issue - I'll need your help there, I think. #210 is baffling - the chain is complex, true, but the scaling obviously happens after application of the tee operator, and [it doesn't look to me as if the tee operator had any code related to resolution](https://github.com/GStreamer/gstreamer/blob/d108ed2251f1f402755d862e83ab0944c552f5c6/plugins/elements/gsttee.c). Maybe it suffices to change the initialization order?

**Comment by severin-lemaignan** [comment]
I've tested the branch, seems to work fine for me. I could not reproduce the flickering either...

**Comment by AndreasBilke** [comment]
Fixed with #251
